### DISCUSSION TITLE
Why does base_dir affect the executed command instead of just what goes into the hash?

### DISCUSSION THREAD

**Discussion by starball5** [ccache/ccache#1487] [Questions & answers]
The following warning is in the docs for `base_dir`:

> Rewriting absolute paths to relative is kind of a brittle hack. It works OK in many cases, but there might be cases where things break. One known issue is that absolute paths are not reproduced in dependency files, which can mess up dependency detection in tools like Make and Ninja. If possible, use relative paths in the first place instead instead of using base_dir.

I use CMake, and [CMake outputs absolute paths](https://gitlab.kitware.com/cmake/community/-/wikis/FAQ#why-does-cmake-use-full-paths-or-can-i-copy-my-build-tree). "Using relative paths in the first place" is [not an option for me](https://cmake.org/cmake/help/latest/variable/CMAKE_USE_RELATIVE_PATHS.html). I use the Ninja generator mainly, but I'd like my project to work with other generators as well.

The thought crossed my mind- If (and maybe I've understood this wrong) `base_dir` is to change what gets hashed to allow for "sharing" cache entries, why not just do the path rewriting for calculating hashes, and leave the paths untouched when actually performing compilation?

I'm not very familiar with ccache (both using it and its internals), so I assume I'm missing something. What am I missing?

**Comment by jrosdahl** [comment]
If absolute paths would be used for the actual compilation command then paths will be incorrect for compilations in other absolute directories. Example:

```console
$ cd /build/alpha
$ echo '#warning example' >example.c
$ ccache base_dir=/build cc -c /build/alpha/example.c
/build/alpha/example.c:1:2: warning: #warning example [-Wcpp]
    1 | #warning example
      |  ^~~~~~~
$ cd /build/bravo
$ echo '#warning example' >example.c
$ ccache base_dir=/build cc -c /build/bravo/example.c
/build/alpha/example.c:1:2: warning: #warning example [-Wcpp]
    1 | #warning example
      |  ^~~~~~~
```

Notice `alpha` in the output instead of `bravo` in the second case. This will happen not only for warning texts but also for strings in the object file (in debug symbols and from expansion of `__FILE__`) and for paths in the dependency file.

By rewriting the actual compiler command to use a relative path this will happen:

```console
$ cd /build/alpha
$ echo '#warning example' >example.c
$ ccache base_dir=/build cc -c /build/alpha/example.c
example.c:1:2: warning: #warning example [-Wcpp]
    1 | #warning example
      |  ^~~~~~~
$ cd /build/bravo
$ echo '#warning example' >example.c
$ ccache base_dir=/build cc -c /build/bravo/example.c
example.c:1:2: warning: #warning example [-Wcpp]
    1 | #warning example
      |  ^~~~~~~
```

**Reply by starball5** [reply]
for compilers that support output path remapping options like `-fprefix-map` (and projects that use those compilers with those options), would there be any other issues with ccache still passing absolute paths to the compiler?

**Reply by jrosdahl** [reply]
Yes, absolute paths would be incorrect for instance in dependency files.

**Reply by starball5** [reply]
Alright thanks for your responses, Joel!
